pkg/comp-functions/functions/vshnnextcloud/deploy.go

```diff
@@ -0,0 +1,395 @@
+
+	comp := &vshnv1.VSHNNextcloud{}
+	err := svc.GetObservedComposite(comp)
+	if err != nil {
+		return runtime.NewFatalResult(fmt.Errorf("cannot get composite: %w", err))
+	}
+
+	svc.Log.Info("Bootstrapping instance namespace and rbac rules")
+	err = common.BootstrapInstanceNs(ctx, comp, comp.GetServiceName(), comp.GetName()+"-instanceNs", svc)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot bootstrap instance namespace: %s", err))
+	}
+
+	if !comp.Spec.Parameters.Service.DefaultInternalDB {
+		svc.Log.Info("Adding postgresql instance")
+		err = addPostgreSQL(svc, comp)
+		if err != nil {
+			return runtime.NewWarningResult(fmt.Sprintf("cannot create postgresql instance: %s", err))
+		}
+
+		svc.Log.Info("Checking readiness of cluster")
+
+		resourceCDMap := map[string][]string{
+			comp.GetName() + pgInstanceNameSuffix: {
+				vshnpostgres.PostgresqlHost,
+				vshnpostgres.PostgresqlPort,
+				vshnpostgres.PostgresqlDb,
+				vshnpostgres.PostgresqlUser,
+				vshnpostgres.PostgresqlPassword,
+			},
+		}
+
+		ready, err := svc.WaitForDependenciesWithConnectionDetails(comp.GetName(), resourceCDMap)
+		if err != nil {
+			// We're returning a fatal here, so in case something is wrong we won't delete anything by mistake.
+			return runtime.NewFatalResult(err)
+		} else if !ready {
+			return runtime.NewWarningResult("postgresql instance not yet ready")
+		}
+	}
+
+	svc.Log.Info("Adding release")
+
+	adminSecret, err := common.AddCredentialsSecret(comp, svc, []string{adminPWSecretField}, common.AddStaticFieldToSecret(map[string]string{adminUserSecretField: "admin"}))
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot generate admin secret: %s", err))
+	}
+
+	cd, err := svc.GetObservedComposedResourceConnectionDetails(adminSecret)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot get observed connection details for nextcloud admin: %s", err))
+	}
+
+	svc.SetConnectionDetail(adminPWConnectionDetailsField, cd[adminPWSecretField])
+	svc.SetConnectionDetail(adminConnectionDetailsField, cd[adminUserSecretField])
+	svc.SetConnectionDetail(hostConnectionDetailsField, []byte(fmt.Sprintf("%s-%s.%s.svc.cluster.local", comp.GetName(), serviceSuffix, comp.GetInstanceNamespace())))
+
+	err = addApacheConfig(svc, comp)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot add configmap for apache: %s", err))
+	}
+
+	err = addRelease(ctx, svc, comp, adminSecret)
+	if err != nil {
+		return runtime.NewWarningResult(fmt.Sprintf("cannot create release: %s", err))
+	}
+
+	return nil
+}
+
+func addPostgreSQL(svc *runtime.ServiceRuntime, comp *vshnv1.VSHNNextcloud) error {
+	// Unfortunately k8up and stackgres backups don't match up very well...
+	// if no daily backup is set we just do the default.
+	retention := 6
+	if comp.Spec.Parameters.Backup.Retention.KeepDaily != 0 {
+		retention = comp.Spec.Parameters.Backup.Retention.KeepDaily
+	}
+
+	configs := map[string]string{
```

Comment:
If I remember correctly, this is specific for Keycloak. We might not need this here.

Reply:
Yes, this was copy&pasted by accident. I've removed it